### PR TITLE
Complement image: propagate SIGTERM to all workers

### PR DESCRIPTION
**Pull request by richvdh**
This should mean that logs from worker processes are flushed before shutdown.

When a test completes, Complement stops the docker container, which means that synapse will receive a SIGTERM. Currently, the `complement_fork_starter` exits immediately (without notifying the worker processes), which means that the workers never get a chance to flush their logs before the whole container is vaped. We can fix this by propagating the SIGTERM to the children.

### CODE REVIEW

File: synapse/app/complement_fork_starter.py [hunk 1]
```diff
@@ -167,13 +179,29 @@ def main() -> None:
     update_proc.join()
     print("===== PREPARED DATABASE =====", file=sys.stderr)
 
+    processes: List[multiprocessing.Process] = []
+
+    # Install signal handlers to propagate signals to all our children, so that they
+    # shut down cleanly. This also inhibits our own exit, but that's good: we want to
+    # wait until the children have exited.
```
Comment (DMRobertson) [inline synapse/app/complement_fork_starter.py:186]: I don't see why our own exit is blocked waiting for our children to exit. Does `os.kill()` block until the child handles the signal? Or does we already have this for free thanks to `process.join()` further below?

**Review by DMRobertson — APPROVED** [review]
LGTM but want to check I've grokked

File: synapse/app/complement_fork_starter.py [hunk 2]
```diff
@@ -167,13 +179,28 @@ def main() -> None:
...
+    def handle_signal(signum: int, frame: Optional[FrameType]) -> None:
+        print(
+            "complement_fork_starter: Caught signal %i. Stopping children." % signum,
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/app/complement_fork_starter.py:189]: ooh fancy.

```suggestion
           f"complement_fork_starter: Caught signal {signum}. Stopping children."
```

[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/app/complement_fork_starter.py:189]: Needs an extra space before the `f` to keep black happy, I think

[the remark this answers is not in the document]
Comment (DMRobertson) [reply synapse/app/complement_fork_starter.py:189]: Err, and the trailing comma

[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/app/complement_fork_starter.py:189]: gah

[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/app/complement_fork_starter.py:189]:
```suggestion
            f"complement_fork_starter: Caught signal {signum}. Stopping children.",
```

File: synapse/app/complement_fork_starter.py [hunk 3]
```diff
@@ -167,13 +179,29 @@ def main() -> None:
...
+            file=sys.stderr,
+        )
+        for p in processes:
+            if p.pid:
+                os.kill(p.pid, signum)
+
+    for sig in (signal.SIGINT, signal.SIGTERM):
+        _original_signal_handlers[sig] = signal.signal(sig, handle_signal)
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/app/complement_fork_starter.py:197]: > I guess the name comes from `signal(2)`?

... which has the same somewhat unintuitive semantics, yes. I guess it made sense 45 years ago.

**Comment by richvdh** [comment]
Complement test failure is https://github.com/matrix-org/complement/issues/478; force-merging

**Pull request merged**